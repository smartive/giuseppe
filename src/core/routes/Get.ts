import 'reflect-metadata';

import { RequestHandler } from 'express';

import { Giuseppe } from '../..';
import { HttpMethod } from '../../routes/RouteDefinition';
import { Callable, FunctionMethodDecorator, GiuseppeBaseRoute } from './GiuseppeBaseRoute';

/**
 * Route decorator. Creates a route definition that reacts to a get request.
 * Can define one or multiple middlewares that are registered for that route. Also, a route name can be specified
 * that creates the url for express.
 *
 * @export
 * @param {(string | RequestHandler)} [routeOrMiddleware] Either a string that represents the url for this route, or
 *                                                        an optional middleware if no specific route is needed.
 * @param {...RequestHandler[]} middlewares Other middlewares that are used for this route.
 * @returns {FunctionMethodDecorator}
 */
export function Get(routeOrMiddleware?: string | RequestHandler, ...middlewares: RequestHandler[]): FunctionMethodDecorator {
  const route = routeOrMiddleware && typeof routeOrMiddleware === 'string' ? routeOrMiddleware : '';
  if (routeOrMiddleware && typeof routeOrMiddleware === 'function') {
      middlewares.unshift(routeOrMiddleware);
    }
  return (target: Object, _: string | symbol, descriptor: TypedPropertyDescriptor<Callable>) => {
      if (!descriptor.value) {
          throw new TypeError(`Function is undefined in route ${route}`);
        }
      Giuseppe.registrar.registerRoute(target, new GiuseppeGetRoute(descriptor.value, route, middlewares));
    };
}

/**
 * Default core get route of giuseppe. Reacts to http.GET.
 *
 * @export
 * @class GiuseppeGetRoute
 * @extends {GiuseppeBaseRoute}
 */
export class GiuseppeGetRoute extends GiuseppeBaseRoute {
  constructor(
        routeFunction: Function,
        route: string = '',
        middlewares: RequestHandler[] = [],
    ) {
      super(HttpMethod.get, routeFunction, route, middlewares);
    }
}
