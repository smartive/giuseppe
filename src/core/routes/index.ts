export { Route } from './GiuseppeBasicRoute';
export { Get } from './Get';
export { Post } from './Post';
