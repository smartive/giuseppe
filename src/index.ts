import 'reflect-metadata';

// Basic types and interfaces for giuseppe
export { Giuseppe } from './Giuseppe';
export * from './GiuseppePlugin';
export * from './controller/ControllerDefinition';
export * from './controller/LoadingOptions';
export { ErrorHandler, ErrorHandlerFunction } from './errors';
export { ParameterDefinition } from './parameter/ParameterDefinition';
export { GiuseppeRoute } from './routes/GiuseppeRoute';
export { ReturnType } from './routes/ReturnType';
export { HttpMethod, RouteDefinition } from './routes/RouteDefinition';
export { RouteModificator } from './routes/RouteModificator';
export * from './utilities/ControllerMetadata';

// Core giuseppe functionallity
export { Controller } from './core/controller/GiuseppeApiController';

export { Body } from './core/parameters/Body';
export { Cookie } from './core/parameters/Cookie';
export { Header } from './core/parameters/Header';
export { isArray, isNumber, isString, ParameterFactory, ParameterValidator, Validator } from './core/parameters/ParameterAdditions';
export { Query } from './core/parameters/Query';
export { UrlParam } from './core/parameters/UrlParam';

export { Delete } from './core/routes/Delete';
export { Get } from './core/routes/Get';
export { Route } from './core/routes/GiuseppeBaseRoute';
export { Head } from './core/routes/Head';
export { Post } from './core/routes/Post';
export { Put } from './core/routes/Put';
