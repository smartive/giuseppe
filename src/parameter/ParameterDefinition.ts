import { Request, Response } from 'express';

/**
 * Reflect metadata key for parameter definitions.
 *
 * @type {string}
 */
export const PARAMETER_DEFINITION_KEY: string = 'giuseppe:ParameterDefinition';

/**
 * Parameter definition. Contains all relevant information about a registered route parameter.
 *
 * @export
 * @interface ParameterDefinition
 */
export interface ParameterDefinition {
    /**
     * The name of the parameter.
     *
     * @type {string}
     * @memberof ParameterDefinition
     */
  readonly name: string;

    /**
     * The type of the parameter. This can be any usable typescript type (class or primitive js type).
     *
     * @type {Function}
     * @memberof ParameterDefinition
     */
  readonly type: Function;

    /**
     * Index that defines the position in the function. Since functions can define non injected values,
     * this index ensures the correct position in the route function.
     *
     * @type {number}
     * @memberof ParameterDefinition
     */
  readonly index: number;

    /**
     * Defines if a parameter definition can handle the response for the user. If this value is "true", giuseppe
     * will not proceed with the return value handlers. Instead giuseppe will assume, that you handle the response
     * on your own in your route function.
     *
     * @type {boolean}
     * @memberof ParameterDefinition
     */
  readonly canHandleResponse: boolean;

    /**
     * Get the actual value of the parameter. Extract the value from the request or the response of express.
     * (e.g. a header parameter would return the header value of the request via "return request.get(this.name)")
     *
     * @param {Request} request
     * @param {Response} response
     * @returns {*}
     * @memberof ParameterDefinition
     */
  getValue(request: Request, response: Response): any;
}
