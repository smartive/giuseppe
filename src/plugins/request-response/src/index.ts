export * from './GiuseppeReqResPlugin';
export * from './Req';
export * from './Res';
