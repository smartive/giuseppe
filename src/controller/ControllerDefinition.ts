import { RequestHandler } from 'express';

import { GiuseppeRoute } from '../routes/GiuseppeRoute';

/**
 * Interface for a controller definition. Contains the targeted function (class) and containing middlewares.
 * The definition must be able to create the routes that are registered within the controller.
 *
 * @export
 * @interface ControllerDefinition
 */
export interface ControllerDefinition {
  /**
   * The target class that is decorated with the controller decorator.
   *
   * @type {Function}
   * @memberof ControllerDefinition
   */
  readonly ctrlTarget: Function;

  /**
   * A collection of middlewares that is contained in the given controller.
   * Middleware priority is: Controller -> Route (they are concatenated).
   *
   * @type {RequestHandler[]}
   * @memberof ControllerDefinition
   */
  readonly middlewares: RequestHandler[];

  /**
   * A function that creates the routes for a given controller. Those routes are then modificated and registered.
   *
   * @param {string} baseUrl
   * @returns {GiuseppeRoute[]}
   * @memberof ControllerDefinition
   */
  createRoutes(baseUrl: string): GiuseppeRoute[];
}
