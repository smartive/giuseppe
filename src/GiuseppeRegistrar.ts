import { ControllerDefinition } from './controller/ControllerDefinition';
import { ErrorHandlerFunction } from './errors/ControllerErrorHandler';
import { ParameterDefinition } from './parameter/ParameterDefinition';
import { RouteDefinition } from './routes/RouteDefinition';
import { RouteModificator } from './routes/RouteModificator';
import { ControllerMetadata } from './utilities/ControllerMetadata';

/**
 * Helper class that is used to register giuseppe - stuff to the specific controllers.
 *
 * @export
 * @class GiuseppeRegistrar
 */
export class GiuseppeRegistrar {
    /**
     * List of controller definitions.
     *
     * @type {ControllerDefinition[]}
     * @memberof GiuseppeRegistrar
     */
  public readonly controller: ControllerDefinition[] = [];

    /**
     * Registeres a controller definition within the register (list of controllers).
     *
     * @param {ControllerDefinition} controller
     * @memberof GiuseppeRegistrar
     */
  public registerController(controller: ControllerDefinition): void {
      this.controller.push(controller);
    }

    /**
     * Register a route to a controllers metadata.
     *
     * @param {Object} controller
     * @param {RouteDefinition} route
     * @memberof GiuseppeRegistrar
     */
  public registerRoute(controller: Object, route: RouteDefinition): void {
      const meta = new ControllerMetadata(controller);
      meta.routes().push(route);
    }

    /**
     * Register a modificator for a route of a controller.
     *
     * @param {Object} controller
     * @param {string} routeName
     * @param {RouteModificator} modificator
     * @memberof GiuseppeRegistrar
     */
  public registerRouteModificator(controller: Object, routeName: string, modificator: RouteModificator): void {
      const meta = new ControllerMetadata(controller);
      meta.modificators(routeName).push(modificator);
    }

    /**
     * Register a parameter for a route of a controller.
     *
     * @param {Object} controller
     * @param {string} routeName
     * @param {ParameterDefinition} parameter
     * @memberof GiuseppeRegistrar
     */
  public registerParameter(controller: Object, routeName: string, parameter: ParameterDefinition): void {
      const meta = new ControllerMetadata(controller);
      meta.parameters(routeName).push(parameter);
    }

    /**
     * Register an error handler for a given error type for a controller.
     *
     * @param {Object} controller
     * @param {ErrorHandlerFunction<Error>} handler
     * @param {Function[]} errors
     * @memberof GiuseppeRegistrar
     */
  public registerErrorHandler(controller: Object, handler: ErrorHandlerFunction<Error>, errors: Function[]): void {
      const meta = new ControllerMetadata(controller);
      for (const error of errors) {
          meta.errorHandler().addHandler(handler, error);
        }
    }
}
