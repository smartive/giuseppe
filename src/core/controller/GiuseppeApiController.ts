import { ControllerDefinition } from '../../controller/ControllerDefinition';
import { Giuseppe } from '../../Giuseppe';
import { RequestHandler, Router } from 'express';

export function Controller(routePrefix: string = '', ...middlewares: RequestHandler[]): ClassDecorator {
    return (ctrl: Function) => Giuseppe.registrar.registerController(new GiuseppeApiController(ctrl, routePrefix, middlewares));
}

export class GiuseppeApiController implements ControllerDefinition {
    constructor(
        public readonly ctrlTarget: Function,
        public readonly routePrefix: string = '',
        public readonly middlewares: RequestHandler[] = [],
    ) { }

    public register(baseUrl: string, router: Router): void {
        throw new Error('Not implemented yet.' + baseUrl + router);
    }
}

// /**
//  * Controller decorator; decorates a class to be a rest api controller. A controller registers itself to an
//  * expressJS router when "registerControllers" is called.
//  *
//  * @param {string} [routePrefix] - Prefix for the whole controller. This path is added to all routes.
//  * @param {RequestHandler[]} [middlewares] - Middleware funct
// ions for the controller to be executed before the routing functions.
//  * @returns {(Function) => void} - Decorator for the controller class.
//  */
// export function Controller(routePrefix?: string, ...middlewares: RequestHandler[]) {
//     return (controller: any) => {
//         IocContainer.get<Registrar>(IoCSymbols.registrar)
// .addController(new ControllerRegistration(controller, routePrefix, middlewares));
//     };
// }
