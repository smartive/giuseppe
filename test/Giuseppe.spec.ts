import { IRouterMatcher, Router } from 'express';
import 'reflect-metadata';
import { Giuseppe, Post, RouteRegisterInformation } from '../src';
import { ControllerDefinition } from '../src/controller/ControllerDefinition';
import { Controller } from '../src/core/controller/GiuseppeApiController';
import { Get } from '../src/core/routes/Get';
import { GiuseppeRegistrar } from '../src/GiuseppeRegistrar';
import { ParameterDefinition } from '../src/parameter/ParameterDefinition';
import { RouteDefinition } from '../src/routes/RouteDefinition';
import { RouteModificator } from '../src/routes/RouteModificator';
import { ControllerMetadata } from '../src/utilities/ControllerMetadata';

describe('Giuseppe', () => {
  describe('Giuseppe registrar', () => {
    afterEach(() => {
      (Giuseppe as any).registrar.controller = [];
    });

    it("should add a controller to it'self", () => {
      class Ctrl {}

      const ctrl: ControllerDefinition = {
        middlewares: [],
        ctrlTarget: Ctrl,
        createRoutes: () => [],
      };

      Giuseppe.registrar.registerController(ctrl);

      expect(Giuseppe.registrar.controller[0]).toBe(ctrl);
    });

    it('should add a route to the controllers metadata', () => {
      class Ctrl {}

      const route: RouteDefinition = {
        middlewares: [],
        httpMethod: 0,
        name: 'name',
        route: 'route',
        routeFunction: () => {},
        createRoutes: () => [],
      };

      Giuseppe.registrar.registerRoute(Ctrl, route);

      const meta = new ControllerMetadata(Ctrl);

      expect(meta.routes().length).toBe(1);
      expect(meta.routes()[0]).toBe(route);
    });

    it('should add a modificator to the controllers metadata', () => {
      class Ctrl {
        public func(): void {}
      }

      const mod: RouteModificator = {
        modifyRoute: () => [],
      };

      Giuseppe.registrar.registerRouteModificator(Ctrl, 'func', mod);

      const meta = new ControllerMetadata(Ctrl);

      expect(meta.modificators('func').length).toBe(1);
      expect(meta.modificators('func')[0]).toBe(mod);
    });

    it('should add a parameter to a route metadata', () => {
      class Ctrl {
        public func(): void {}
      }

      const param: ParameterDefinition = {
        canHandleResponse: true,
        getValue: () => '',
        index: 0,
        name: 'name',
        type: String,
      };

      Giuseppe.registrar.registerParameter(Ctrl, 'func', param);

      const meta = new ControllerMetadata(Ctrl);

      expect(meta.parameters('func').length).toBe(1);
      expect(meta.parameters('func')[0]).toBe(param);
    });

    it('should add an error handler to a controller metadata', () => {
      class Ctrl {}

      const fn = () => {};

      Giuseppe.registrar.registerErrorHandler(Ctrl, fn, [Error]);

      const meta = new ControllerMetadata(Ctrl);

      expect((meta.errorHandler() as any).handlers.Error).toBe(fn);
    });
  });

  describe('Core', () => {
    describe('plugin operations', () => {
      let giuseppe: Giuseppe;
      let plugin: any;

      beforeEach(() => {
        giuseppe = new Giuseppe();
        (giuseppe as any).plugins = [];
        plugin = {
          name: 'plugin',
          initialize: () => {},
          controllerDefinitions: null,
          parameterDefinitions: null,
          routeDefinitions: null,
          routeModificators: null,
          returnTypeHandler: null,
        };
      });

      describe('returnTypes()', () => {
        it('should return the definitions', () => {
          expect((giuseppe as any)._returnTypes).toBeNull();

          plugin.returnTypeHandler = [
            {
              getHeaders: () => ({}),
              getStatus: () => 0,
              getValue: () => '',
              type: 'default',
            },
          ];
          giuseppe.registerPlugin(plugin);

          expect((giuseppe as any).returnTypes.length).toBe(1);
        });
      });

      describe('pluginController()', () => {
        it('should return the definitions', () => {
          expect((giuseppe as any)._pluginController).toBeNull();

          plugin.controllerDefinitions = [() => {}];
          giuseppe.registerPlugin(plugin);

          expect((giuseppe as any).pluginController.length).toBe(1);
        });
      });

      describe('pluginRoutes()', () => {
        it('should return the definitions', () => {
          expect((giuseppe as any)._pluginRoutes).toBeNull();

          plugin.routeDefinitions = [() => {}];
          giuseppe.registerPlugin(plugin);

          expect((giuseppe as any).pluginRoutes.length).toBe(1);
        });
      });

      describe('pluginRouteModificators()', () => {
        it('should return the definitions', () => {
          expect((giuseppe as any)._pluginRouteModificators).toBeNull();

          plugin.routeModificators = [() => {}];
          giuseppe.registerPlugin(plugin);

          expect((giuseppe as any).pluginRouteModificators.length).toBe(1);
        });
      });

      describe('pluginParameters()', () => {
        it('should return the definitions', () => {
          expect((giuseppe as any)._pluginParameters).toBeNull();

          plugin.parameterDefinitions = [() => {}];
          giuseppe.registerPlugin(plugin);

          expect((giuseppe as any).pluginParameters.length).toBe(1);
        });
      });

      describe('registerPlugin()', () => {
        it('should reset the plugin list', () => {
          (giuseppe as any).returnTypes;
          (giuseppe as any).pluginController;
          (giuseppe as any).pluginRoutes;
          (giuseppe as any).pluginRouteModificators;
          (giuseppe as any).pluginParameters;

          expect((giuseppe as any).returnTypes.length).toBe(0);
          expect((giuseppe as any).pluginController.length).toBe(0);
          expect((giuseppe as any).pluginRoutes.length).toBe(0);
          expect((giuseppe as any).pluginRouteModificators.length).toBe(0);
          expect((giuseppe as any).pluginParameters.length).toBe(0);

          giuseppe.registerPlugin(plugin);

          expect((giuseppe as any)._returnTypes).toBeNull();
          expect((giuseppe as any)._pluginController).toBeNull();
          expect((giuseppe as any)._pluginRoutes).toBeNull();
          expect((giuseppe as any)._pluginRouteModificators).toBeNull();
          expect((giuseppe as any)._pluginParameters).toBeNull();
        });
      });
    });

    describe('configureRouter()', () => {
      let giuseppe: Giuseppe;
      let spy: jest.SpyInstance<IRouterMatcher<Router>> | null;

      beforeEach(() => {
        giuseppe = new Giuseppe();
      });

      afterEach(() => {
        (Giuseppe as any).registrar = new GiuseppeRegistrar();

        if (spy) {
          spy.mockReset();
          spy.mockRestore();
          spy = null;
        }
      });

      it('should use the correct baseUrl', () => {
        @Controller()
        class Ctrl {
          @Get('getFunc')
          public func(): void {}
        }

        spy = jest.spyOn(giuseppe.router, 'get');

        giuseppe.configureRouter('baseUrl');

        expect(spy).toHaveBeenCalled();
        expect(spy.mock.calls[0][0]).toBe('/baseUrl/getFunc');
      });

      it('should use a controller prefix correctly', () => {
        @Controller('prefix')
        class Ctrl {
          @Get('getFunc')
          public func(): void {}
        }

        spy = jest.spyOn(giuseppe.router, 'get');

        giuseppe.configureRouter();

        expect(spy).toHaveBeenCalled();
        expect(spy.mock.calls[0][0]).toBe('/prefix/getFunc');
      });

      it('should use baseUrl and controller prefix correctly', () => {
        @Controller('prefix')
        class Ctrl {
          @Get('getFunc')
          public func(): void {}
        }

        spy = jest.spyOn(giuseppe.router, 'get');

        giuseppe.configureRouter('baseUrl');

        expect(spy).toHaveBeenCalled();
        expect(spy.mock.calls[0][0]).toBe('/baseUrl/prefix/getFunc');
      });

      it('should register a higher segmented route first', () => {
        @Controller()
        class Ctrl {
          @Get('/this/has/four/segments')
          public func(): void {}

          @Get('/this/has/five/segments/yo')
          public func2(): void {}
        }

        spy = jest.spyOn(giuseppe.router, 'get');

        giuseppe.configureRouter();

        expect(spy.mock.calls[0][0]).toBe('/this/has/five/segments/yo');
        expect(spy.mock.calls[1][0]).toBe('/this/has/four/segments');
      });

      it('should register a lower parameterized route first', () => {
        @Controller()
        class Ctrl {
          @Get('/url/:id/:type')
          public func(): void {}

          @Get('/this/:id/notparam')
          public func2(): void {}

          @Get('/this/no/hotdog')
          public func3(): void {}
        }

        spy = jest.spyOn(giuseppe.router, 'get');

        giuseppe.configureRouter();

        expect(spy.mock.calls[0][0]).toBe('/this/no/hotdog');
        expect(spy.mock.calls[1][0]).toBe('/this/:id/notparam');
        expect(spy.mock.calls[2][0]).toBe('/url/:id/:type');
      });

      it('should register a lower wildcarded route first', () => {
        @Controller()
        class Ctrl {
          @Get('func1/*/2/*')
          public func(): void {}

          @Get('func1/foo/*/bar')
          public func2(): void {}
        }

        spy = jest.spyOn(giuseppe.router, 'get');

        giuseppe.configureRouter();

        expect(spy.mock.calls[0][0]).toBe('/func1/foo/*/bar');
        expect(spy.mock.calls[1][0]).toBe('/func1/*/2/*');
      });

      it('should not contain double slashes in routes', () => {
        @Controller('/')
        class Ctrl {
          @Get('/foo//bar')
          public func(): void {}
        }

        spy = jest.spyOn(giuseppe.router, 'get');

        giuseppe.configureRouter('/');

        expect(spy.mock.calls[0][0]).toBe('/foo/bar');
      });

      it('should throw on a duplicate route', () => {
        @Controller()
        class Ctrl {
          @Get('getFunc')
          public func(): void {}

          @Get('getFunc')
          public func2(): void {}
        }

        const fn = () => giuseppe.configureRouter('baseUrl');

        expect(fn).toThrowErrorMatchingSnapshot();
      });

      it('should not throw on a similar route in a different controller', () => {
        @Controller('a')
        class CtrlA {
          @Get('getFunc')
          public func(): void {}
        }

        @Controller('b')
        class CtrlB {
          @Get('getFunc')
          public func(): void {}
        }

        const fn = () => giuseppe.configureRouter('baseUrl');

        expect(fn).not.toThrow();
      });

      it('should order the urls correctly (all cases)', () => {
        @Controller()
        class Ctrl {
          @Get('*')
          public funcGet5(): void {}

          @Get('url/:two/:params')
          public funcGet3(): void {}

          @Get('url/:two/*')
          public funcGet7(): void {}

          @Get('url/foo/bar')
          public funcGet9(): void {}

          @Get('url/*/*')
          public funcGet8(): void {}

          @Get('url/with/many/segments/there')
          public funcGet(): void {}

          @Get('url/with/:many/segments/there')
          public funcGet6(): void {}

          @Get('url/with/*/segments/there')
          public funcGet4(): void {}

          @Get('url/*/:param')
          public funcGet11(): void {}

          @Get('url/foo/:param')
          public funcGet12(): void {}

          @Get('url/*/param')
          public funcGet10(): void {}

          @Get('url/:one/param')
          public funcGet2(): void {}
        }

        spy = jest.spyOn(giuseppe.router, 'get');

        giuseppe.configureRouter();

        expect(spy.mock.calls).toMatchSnapshot();
      });
    });
  });

  describe('createRouteWrapper()', () => {
    it('should use the same controller instance for routes', () => {
      const giusi = new Giuseppe();
      const original = (giusi as any).createRouteWrapper as Function;
      const spy = jest.fn((routeInfo: RouteRegisterInformation) => original.apply(giusi, [routeInfo]));
      (giusi as any).createRouteWrapper = spy;

      @Controller('foo')
      class Ctrl {
        @Get('bar')
        public func1(): void {}

        @Post('baz')
        public async func2(p1: string): Promise<void> {}
      }

      @Controller('bar')
      class Ctrl2 {
        @Get('bar')
        public func1(): void {}
      }

      giusi.configureRouter();

      expect(spy.mock.calls[0][0].instance).toBe(spy.mock.calls[1][0].instance);
      expect(spy.mock.calls[0][0].instance).not.toBe(spy.mock.calls[2][0].instance);
    });
  });
});
