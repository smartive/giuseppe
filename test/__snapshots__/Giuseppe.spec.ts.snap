// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Giuseppe Core configureRouter() should order the urls correctly (all cases) 1`] = `
Array [
  Array [
    "/url/with/many/segments/there",
    [Function],
  ],
  Array [
    "/url/with/:many/segments/there",
    [Function],
  ],
  Array [
    "/url/with/*/segments/there",
    [Function],
  ],
  Array [
    "/url/foo/bar",
    [Function],
  ],
  Array [
    "/url/foo/:param",
    [Function],
  ],
  Array [
    "/url/:one/param",
    [Function],
  ],
  Array [
    "/url/:two/:params",
    [Function],
  ],
  Array [
    "/url/*/param",
    [Function],
  ],
  Array [
    "/url/*/:param",
    [Function],
  ],
  Array [
    "/url/:two/*",
    [Function],
  ],
  Array [
    "/url/*/*",
    [Function],
  ],
  Array [
    "/*",
    [Function],
  ],
]
`;

exports[`Giuseppe Core configureRouter() should throw on a duplicate route 1`] = `"A route with the ID 'get_baseUrl/getFunc' (method name: 'func2', url: 'baseUrl/getFunc') is already registered."`;
